WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Grouped Products/ProductListMultiSelectorSearchUICommand.swift

```diff
@@ -92,13 +95,43 @@ final class ProductListMultiSelectorSearchUICommand: SearchUICommand {
     }
 }
 
+extension ProductListMultiSelectorSearchUICommand {
+    /// Configures the given presenting view controller to present unsaved changes alert when the user attempts to dismiss the search modal with
+    /// unsaved changes.
+    func configurePresentingViewControllerForDiscardChangesAlert(presentingViewController: UIViewController) {
+        self.presentingViewController = presentingViewController
+        presentingViewController.presentationController?.delegate = self
+    }
+}
+
+extension ProductListMultiSelectorSearchUICommand: UIAdaptivePresentationControllerDelegate {
+    func presentationControllerDidAttemptToDismiss(_ presentationController: UIPresentationController) {
+        guard let presentingViewController = presentingViewController else {
+            return
+        }
+        guard hasUnsavedChanges() else {
+            presentingViewController.dismiss(animated: true)
+            return
+        }
+        UIAlertController.presentDiscardChangesActionSheet(viewController: presentingViewController,
+                                                           onDiscard: { [weak self] in
+                                                            self?.presentingViewController?.dismiss(animated: true)
+        })
+    }
```

Comment:
While I was testing this something in the drag gesture felt off when dismissing a view controller that does not have any pending changes. And it looks like this is because we are prohibiting all gestures from the start and then deciding what to do.

In order to have the same experience when dismissing a `non-dirty` view we could do this

```suggestion
    func presentationControllerShouldDismiss(_ presentationController: UIPresentationController) -> Bool {
        return !hasUnsavedChanges()
    }

    func presentationControllerDidAttemptToDismiss(_ presentationController: UIPresentationController) {
        UIAlertController.presentDiscardChangesActionSheet(viewController: presentationController.presentedViewController, onDiscard: {
            presentationController.presentedViewController.dismiss(animated: true)
        })
    }
```

Here the `presentationControllerShouldDismiss` method controls if the controller should be dismissed or not.
and `presentationControllerDidAttemptToDismiss` is only called when the user attempted to dismiss it when we didn't allow it.

Reply:
oooh thanks for the suggestions! this feels like a much better experience, updated in https://github.com/woocommerce/woocommerce-ios/pull/2976/commits/2ad8c7d871dabfe25364030026e3159f250473de and https://github.com/woocommerce/woocommerce-ios/pull/2976/commits/ce115266d1acd8327ced3274576a8247b1658424